### PR TITLE
Add: rspec fixtures

### PR DESCRIPTION
**Pull request by benkoshy**
* Incomplete - spec files still refuse to copy  to the expected destination. no idea why. embarassingly, i spent a fair amount of time without getting anywhere.  Here is the latest.

### CODE REVIEW

File: lib/generators/rodauth/install_generator.rb [hunk 1]
```diff
@@ -5,6 +5,9 @@ module Rodauth
   module Rails
     module Generators
       class InstallGenerator < ::Rails::Generators::Base
+        # source_root "#{__dir__}/templates"
+        # destination_root = "#{__dir__}/templates"
+
```
Comment (janko) [inline lib/generators/rodauth/install_generator.rb:10]: Why comment this out?

File: lib/generators/rodauth/install_generator.rb [hunk 2]
```diff
@@ -71,9 +74,14 @@ def create_mailer
...
         def create_fixtures
-          test_unit_options = ::Rails.application.config.generators.options[:test_unit]
-          if test_unit_options[:fixture] && test_unit_options[:fixture_replacement].nil?
-            template "app/test/fixtures/accounts.yml"
+          test_unit_options = ::Rails.application.config.generators.options[:test_unit]          
+          if test_unit_options[:fixture] && test_unit_options[:fixture_replacement].nil?                                    
+            if defined?(RSpec::Rails)              
+              copy_file "app/test/fixtures/accounts.yml" "app/spec/fixtures/accounts.yml"               
+            else
+              template "app/test/fixtures/accounts.yml"
```
Comment (janko) [inline lib/generators/rodauth/install_generator.rb:82]: I noticed this only know, but it seems to be copying to `app/test/` instead of `test/`.

File: lib/generators/rodauth/install_generator.rb [hunk 3]
```diff
@@ -71,9 +74,14 @@ def create_mailer
         end
 
         def create_fixtures
-          test_unit_options = ::Rails.application.config.generators.options[:test_unit]
-          if test_unit_options[:fixture] && test_unit_options[:fixture_replacement].nil?
-            template "app/test/fixtures/accounts.yml"
+          test_unit_options = ::Rails.application.config.generators.options[:test_unit]          
+          if test_unit_options[:fixture] && test_unit_options[:fixture_replacement].nil?                                    
+            if defined?(RSpec::Rails)
```
Comment (janko) [inline lib/generators/rodauth/install_generator.rb:79]: Did you try checking the corrected generator option instead? Seems like it would be more robust, and testable.

**Comment by janko** [comment]
Is the issue that tests are not passing, or that it's not working in an actual app using RSpec?

**Review by janko — COMMENTED** [review]
Is the issue that tests are not passing, or that it's not working in an actual app using RSpec?

File: lib/generators/rodauth/install_generator.rb (hunk 1 above)
Comment (benkoshy) [reply lib/generators/rodauth/install_generator.rb]: > Why comment this out?
I wasn't sure exactly of what they were doing, and to alert you to their existence.

File: lib/generators/rodauth/install_generator.rb (hunk 2 above)
Comment (benkoshy) [reply lib/generators/rodauth/install_generator.rb]: good find. i have rectified.

**Comment by benkoshy** [comment]
The culprit was a missing comma! 

The only thing remaining is to correctly stub the Rails.config object and to test for that. if you have a pointer on how to efficiently do this (just the general idea will more than suffice - to save you time etc.) pls LMK and I can finish off the rest..

**Comment by janko** [comment]
I don't think the RSpec branch is worth testing, I would just not test it. I also skipped tests for disabled fixtures, because I didn't want to deal with global mutable state that is `config.generators`.

**Comment by benkoshy** [comment]
I manually tested for spec using the rodauth-rails generator - looks good:

image

File: lib/generators/rodauth/install_generator.rb [hunk 4]
```diff
@@ -5,6 +5,9 @@ module Rodauth
   module Rails
     module Generators
       class InstallGenerator < ::Rails::Generators::Base
+        source_root "#{__dir__}/templates"
+        destination_root = "#{__dir__}/"
+
```
Comment (janko) [inline lib/generators/rodauth/install_generator.rb:10]: Why is this here?

File: lib/generators/rodauth/install_generator.rb [hunk 5]
```diff
@@ -71,9 +74,13 @@ def create_mailer
...
         def create_fixtures
-          test_unit_options = ::Rails.application.config.generators.options[:test_unit]
-          if test_unit_options[:fixture] && test_unit_options[:fixture_replacement].nil?
-            template "app/test/fixtures/accounts.yml"
+          test_unit_options = ::Rails.application.config.generators.options[:test_unit]          
+          if test_unit_options[:fixture] && test_unit_options[:fixture_replacement].nil?                                    
+            if ::Rails.application.config.generators.options[:rails][:test_framework] == :rspec            
+              copy_file "app/test/fixtures/accounts.yml", "spec/fixtures/accounts.yml"               
+            else              
+              template "app/test/fixtures/accounts.yml", "test/fixtures/accounts.yml"
```
Comment (janko) [inline lib/generators/rodauth/install_generator.rb:82]: Should the source fixture continue to live in `templates/app/test/`, or wouldn't it make more sense to move it to `templates/test/` as well?

File: lib/generators/rodauth/install_generator.rb [hunk 6]
```diff
@@ -71,9 +74,13 @@ def create_mailer
         end
 
         def create_fixtures
-          test_unit_options = ::Rails.application.config.generators.options[:test_unit]
-          if test_unit_options[:fixture] && test_unit_options[:fixture_replacement].nil?
-            template "app/test/fixtures/accounts.yml"
+          test_unit_options = ::Rails.application.config.generators.options[:test_unit]          
+          if test_unit_options[:fixture] && test_unit_options[:fixture_replacement].nil?                                    
+            if ::Rails.application.config.generators.options[:rails][:test_framework] == :rspec            
+              copy_file "app/test/fixtures/accounts.yml", "spec/fixtures/accounts.yml"
```
Comment (janko) [inline lib/generators/rodauth/install_generator.rb:80]: Why are we using `copy_file` for RSpec branch, and `template` for minitest branch?

File: test/generators/install_generator_test.rb [hunk 7]
```diff
@@ -101,9 +101,9 @@ class InstallGeneratorTest < Rails::Generators::TestCase
     end
   end
 
-  test "fixture" do
+  test "fixtures" do
```
Comment (janko) [inline test/generators/install_generator_test.rb:104]: Extra whitespace ✂️

File: test/generators/install_generator_test.rb [hunk 8]
```diff
@@ -101,9 +101,9 @@ class InstallGeneratorTest < Rails::Generators::TestCase
     end
   end
 
-  test "fixture" do
+  test "fixtures" do      	     
     run_generator
-
-    assert_file "app/test/fixtures/accounts.yml", /<%= BCrypt::Password/
+    assert_file "test/fixtures/accounts.yml", /BCrypt::Password/ # No tests for the rspec branch
```
Comment (janko) [inline test/generators/install_generator_test.rb:106]: Asserting the presence of the ERB tag `<%= ` was intentional, why remove it?

**Review by janko — CHANGES_REQUESTED** [review]

File: lib/generators/rodauth/install_generator.rb (hunk 5 above)
Comment (benkoshy) [reply lib/generators/rodauth/install_generator.rb:82]: > Should the template fixture continue to live in app/test/, or wouldn't it make more sense to move it to test/ as well?

Depends how you view the situation:  when I think `rodauth-rails/test` I think about tests for the `rodauth-rails` library itself, rather than templates for it. Pls confirm how you want to proceed?

File: lib/generators/rodauth/install_generator.rb (hunk 6 above)
Comment (benkoshy) [reply lib/generators/rodauth/install_generator.rb:80]: Your right. I was using `copy_file` and forgot to switch it over. template would be more accurate. https://rdoc.info/github/wycats/thor/Thor%2FActions:template

File: lib/generators/rodauth/install_generator.rb (hunk 5 above)
Comment (janko) [reply lib/generators/rodauth/install_generator.rb:82]: Note that my original comment was only talking about files within the `lib/generators/rodauth/templates/` directory.

File: test/generators/install_generator_test.rb (hunk 8 above)
Comment (benkoshy) [reply test/generators/install_generator_test.rb:106]: By accident.

File: lib/generators/rodauth/install_generator.rb (hunk 5 above)
Comment (benkoshy) [reply lib/generators/rodauth/install_generator.rb]: I see. So the fixtures, according to my tests in a real life rails app go in this directory:
Rspec Rails app: `spec/fixtures`
Other testing frameworks: `test/fixtures`  

The do not seem to be installing in the `app/tests` directory when a rrails app is being used. proof:

**Rspec**
image
image

**test unit:**
image

Comment (benkoshy) [reply lib/generators/rodauth/install_generator.rb]: Perhaps I am not understanding: please correct me if so?

Comment (janko) [reply lib/generators/rodauth/install_generator.rb]: I'm talking about the templates directory inside rodauth-rails, not where the fixtures get imported into the app (that part was fixed). I don't know how else to explain it. We should not have an `app/test` directory anywhere, neither in rodauth-rails nor in Rails app when running `rodauth:install`.

Comment (benkoshy) [reply lib/generators/rodauth/install_generator.rb]: Ok now I understand! Apologies for my confusion. let me try for your proposal.

Comment (benkoshy) [reply lib/generators/rodauth/install_generator.rb]: https://github.com/janko/rodauth-rails/pull/127/commits/04717782c23534ab2f1b0c2e1451d07591974298 

Pls let me know if anything unresolved. i can squash and submit the final PR.

**Comment by janko** [comment]
The changes look good now, the only thing that remains is removing the unrelated whitespce changes, as those make git blame more difficult to use.

**Comment by benkoshy** [comment]
Should be ready now. Apologies, I made some mistakes.

Thank you for your umpteen and magnanimous patience.

**Review by janko — APPROVED** [review]
Looks good now, thanks!
